dartagnan/src/main/java/com/dat3m/dartagnan/encoding/ProverWithTracker.java

```diff
@@ -135,10 +109,21 @@ private void write(String content) {
         try {
             writer = new FileWriter(file, true);
             PrintWriter printer = new PrintWriter(writer);
-            printer.append(content);
+            printer.append(removeDuplicatedDeclarations(content));
```

Comment:
This cannot work, can it? A single formula won't contain duplicates, it will contain duplicates wrt. the existing file content. I I think you need to traverse the whole file every single time you append something.

Reply:
Notice that I redefined the `removeDuplicatedDeclarations`. Now the class keep tracks of declarations, and each single formula will check if the variables that it is trying to define were already defined by some previous one, and if so, it skips them.

Attached are the smtlib2 files generated from the last commit. They seem correct to me (at least the overall structure makes sense).

[output.zip](https://github.com/user-attachments/files/16796300/output.zip)

Reply:
Oh sry, I didn't see this. Then it is fine (though it still has the same theoretical issue I mentioned already).

Reply:
No. If you have
```
push
declare x
pop
declare x
assert (P(x))
```
then both declaration are valid/necessary, because the former gets popped before the redeclaration. Your procedure would avoid printing the second declaration, generating an invalid smtlib file.
I don't think this happens in our current usage of the solver, so you get away with "ignoring the stack structure".